src/config/config_specification.cpp

```diff
@@ -0,0 +1,173 @@
+#include "config/config_specification.h"
+
+#include <cstring>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <unordered_set>
+#include <variant>
+#include <vector>
+
+#include <boost/algorithm/string/join.hpp>
+#include <boost/core/span.hpp>
+#include <boost/lexical_cast.hpp>
+
+#include "config/config_key_path.h"
+#include "config/source/command_line_arguments.h"
+#include "config/source/environment_variables.h"
+#include "config/source/yaml_file.h"
+#include "silo/common/cons_list.h"
+#include "silo/common/fmt_formatters.h"
+#include "silo/common/panic.h"
+#include "silo/common/string_utils.h"
+
+namespace {
+std::string indent(std::string_view indentation, const std::string& str) {
+   auto lines = silo::splitBy(str, "\n");
+   std::string out{};
+   for (const auto& line : lines) {
+      // Can't do: out.push_back(std::string { indentation });
+      for (const char character : indentation) {
+         out.push_back(character);
+      }
+      for (const char character : line) {
+         out.push_back(character);
+      }
+      out.push_back('\n');
+   }
+   return out;
+}
+}  // namespace
+
+namespace silo::config {
+
+std::optional<ConfigAttributeSpecification> ConfigSpecification::
+   getAttributeSpecificationFromAmbiguousKey(const silo::config::AmbiguousConfigKeyPath& key
+   ) const {
+   for (const auto& field : attribute_specifications) {
+      if (key == AmbiguousConfigKeyPath::from(field.key)) {
+         return field;
+      }
+   }
+   return std::nullopt;
+}
+
+std::optional<ConfigAttributeSpecification> ConfigSpecification::getAttributeSpecification(
+   const silo::config::ConfigKeyPath& key
+) const {
+   auto maybe_result = std::find_if(
+      attribute_specifications.begin(),
+      attribute_specifications.end(),
+      [&](const ConfigAttributeSpecification& attribute_spec) { return attribute_spec.key == key; }
+   );
+   if (maybe_result == attribute_specifications.end()) {
+      return std::nullopt;
+   }
+   return *maybe_result;
+}
```

Comment:
The second one could use the AssociationList wrapper I mentioned (include/silo/common/association_list.h). 

The first might, too, if adding a C++ constructor for AmbiguousConfigKeyPath from ConfigKeyPath (instead of the static `AmbiguousConfigKeyPath::from` method)--I'm not sure whether to think of this as horrible or as typical C++.

Reply:
I think the main difference in implementation here is how we search the right entry in a list: 
* The first methods uses the classic `for (...) { if (match) return; } return nothing` 
* The second one uses the "functional approach" with `find_if`

The only real difference (in terms of logic) is `key == AmbiguousConfigKeyPath::from(field.key)` vs `attribute_spec.key == key`. It would be good if one could see that immediately instead of hiding it behind different list search implementations. I think I have a slight preference for the classic `for if return` approach (less code in C++), but I don't mind as long as they are similar.

Reply:
(Sidenote: I don't mind raw loops - in languages that don't have a proper way to handle functional style they are often easier to read and C++ is one of those languages. But that's a matter of taste and a discussion for another time)